MetricsUploader/include/MetricsUploader/MetricsUploader.h

```diff
@@ -0,0 +1,76 @@
+// Copyright (c) 2020 The Orbit Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#ifndef ORBIT_METRICS_UPLOADER_METRICS_UPLOADER_H_
+#define ORBIT_METRICS_UPLOADER_METRICS_UPLOADER_H_
+
+#ifdef _WIN32
+#include <windows.h>
+#endif  // WIN32
+
+#include <absl/synchronization/mutex.h>
+
+#include <cstdint>
+#include <memory>
+#include <string>
+#include <string_view>
+
+#include "MetricsUploader/Result.h"
+#include "OrbitBase/Result.h"
+#include "orbit_log_event.pb.h"
+
+namespace orbit_metrics_uploader {
+
+constexpr std::string_view kMetricsUploaderClientDllName = "metrics_uploader_client";
+
+// This class is used for sending log events from Orbit. It works only on Windows and only if
+// metrics_uploader_client.dll is available. The types of logs that could be send are defined in
+// orbit_client_protos::OrbitLogEvent::LogEventType.
+//
+// Only one instance of the class is allowed.
+//
+// Usage example:
+//
+// auto metrics_uploader = MetricsUploader::CreateMetricsUploader();
+// if (metrics_uploader.has_value()) {
+//   metrics_uploader.value()->SendLogEvent(...);
+// }
+//
+class MetricsUploader {
+ public:
+  // Create a MetricsUploader instance when is called first time and ErrorMessage otherwise.
+  [[nodiscard]] static ErrorMessageOr<std::unique_ptr<MetricsUploader>> CreateMetricsUploader(
+      const std::string_view client_name = kMetricsUploaderClientDllName);
+
+  // Unload metrics_uploader_client.dll
+  ~MetricsUploader();
+
+  MetricsUploader(const MetricsUploader& other) = delete;
+  MetricsUploader(MetricsUploader&& other) = default;
+  MetricsUploader& operator=(const MetricsUploader& other) = delete;
+  MetricsUploader& operator=(MetricsUploader&& other) = default;
+
+  // Send log events to the server using metrics_uploader.
+  // Returns true on success and false otherwise.
+  bool SendLogEvent(OrbitLogEvent_LogEventType log_event_type);
+
+ private:
+  // Load metrics uploader client library if available and starts metrics uploader client.
+  MetricsUploader(const std::string_view client_name = kMetricsUploaderClientDllName);
+
+  void Reset();
```

Comment:
If I understand everything correctly: The user should call `CreateMetricsUploader` and will only get a result if no instance has been created before. So that's good.

If no instance has been created yet, the CreateMetricsUploader function will create a new `MetricsUploader` instance. But the construction of that instance can fail which is checked, by `is_init_`.

What about moving all the code from the constructor to `CreateMetricsUploader`? This would simplify the design since `Reset won't be needed anymore and `is_init` will only be changed in one place - in `CreateMetricsUploader`

Reply:
I reorganized this part of the code, could you please take another look?